src/json_schema/parsing.rs

```diff
@@ -1,578 +1,629 @@
-    _whitespace_pattern: &str,
-) -> Result<String> {
-    match obj.get("const") {
-        Some(const_value) => match const_value {
-            Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_) => {
-                let json_string = serde_json::to_string(const_value)?;
-                Ok(regex::escape(&json_string))
+    fn parse_enum(&mut self, obj: &serde_json::Map<String, Value>) -> Result<String> {
+        match obj.get("enum") {
+            Some(Value::Array(enum_values)) => {
+                let choices: Result<Vec<String>> = enum_values
+                    .iter()
+                    .map(|choice| match choice {
+                        Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_) => {
+                            let json_string = serde_json::to_string(choice)?;
+                            Ok(regex::escape(&json_string))
+                        }
+                        _ => Err(JsonSchemaParserError::UnsupportedEnumDataType(Box::new(
+                            choice.clone(),
+                        ))),
+                    })
+                    .collect();
+
+                let choices = choices?;
+                Ok(format!(r"({})", choices.join("|")))
             }
-            _ => Err(anyhow!("Unsupported data type in const: {:?}", const_value)),
-        },
-        None => Err(anyhow!("'const' key not found in object")),
+            _ => Err(JsonSchemaParserError::EnumMustBeAnArray),
+        }
+    }
+
+    fn parse_const(&mut self, obj: &serde_json::Map<String, Value>) -> Result<String> {
+        match obj.get("const") {
+            Some(const_value) => match const_value {
+                Value::Null | Value::Bool(_) | Value::Number(_) | Value::String(_) => {
+                    let json_string = serde_json::to_string(const_value)?;
+                    Ok(regex::escape(&json_string))
+                }
+                _ => Err(JsonSchemaParserError::UnsupportedConstDataType(Box::new(
+                    const_value.clone(),
+                ))),
+            },
+            None => Err(JsonSchemaParserError::ConstKeyNotFound),
+        }
     }
-}
 
-pub fn parse_ref(
-    obj: &serde_json::Map<String, Value>,
-    whitespace_pattern: &str,
-    full_schema: &Value,
-) -> Result<String> {
-    let ref_path = obj["$ref"]
-        .as_str()
-        .ok_or_else(|| anyhow!("'$ref' must be a string"))?;
-
-    let parts: Vec<&str> = ref_path.split('#').collect();
-
-    match parts.as_slice() {
-        [fragment] | ["", fragment] => {
-            let path_parts: Vec<&str> = fragment.split('/').filter(|&s| !s.is_empty()).collect();
-            let referenced_schema = resolve_local_ref(full_schema, &path_parts)?;
-            to_regex(referenced_schema, Some(whitespace_pattern), full_schema)
+    fn parse_ref(&mut self, obj: &serde_json::Map<String, Value>) -> Result<String> {
+        if self.recursion_depth > self.max_recursion_depth {
```

Comment:
should this be `>=` given that it's incremented on 285?

Reply:
(if so, you could safely bump the max depth to 4 and keep the same performance)

Reply:
I considered that and decided against it as counterintuitive from usage perspective.

Imagine the case, where I want to set `recursion_depth` to 0, which means literally no recursion, please. So it actually means what it says from both sides:
- `recursion_depth-0` -> no recursion
- `max_recursion_depth=3` -> 3 levels
 
With +1 to set to both, 0 case becomes invalid and we need to keep `-1` in mind to translate both to something meaningful.

Reply:
Fair enough!